src/state-manager/TransactionConsensus.ts

```diff
@@ -1024,6 +1094,565 @@ class TransactionConsenus {
+              nodesToSendTo.add(node)
+            }
+          }
+          if (nodesToSendTo.size > 0) {
+            Comms.sendGossip(
+              'gossip-final-state',
+              readableReq.finalState,
+              null,
+              null,
+              Array.from(nodesToSendTo),
+              false,
+              4,
+              queueEntry.acceptedTx.txId
+            )
+            nestedCountersInstance.countEvent(`processing`, `forwarded final data to storage nodes`)
+          }
+        } catch (e) {
+          console.error(`Error processing poqoDataAndReceipt Binary handler: ${e}`)
+          nestedCountersInstance.countEvent('internal', `${route}-exception`)
+          this.mainLogger.error(`${route}: Exception executing request: ${utils.errorToStringFull(e)}`)
+        } finally {
+          profilerInstance.scopedProfileSectionEnd(route)
+        }
+      },
+    }
+    Comms.registerInternalBinary(
+      poqoDataAndReceiptBinaryHandler.name,
+      poqoDataAndReceiptBinaryHandler.handler
+    )
+
+    Comms.registerInternal(
+      'poqo-data-and-receipt',
+      async (
+        payload: {
+          finalState: { txid: string; stateList: Shardus.WrappedResponse[] }, 
+          receipt: AppliedReceipt2
+        }, 
+        _respond: unknown,
+        _sender: string,
+      ) => {
+        profilerInstance.scopedProfileSectionStart('poqo-data-and-receipt')
+        try {
+          // make sure we have it
+          const queueEntry = this.stateManager.transactionQueue.getQueueEntrySafe(payload.finalState.txid) // , payload.timestamp)
+          //It is okay to ignore this transaction if the txId is not found in the queue.
+          if (queueEntry == null) {
+            //In the past we would enqueue the TX, expecially if syncing but that has been removed.
+            //The normal mechanism of sharing TXs is good enough.
+            nestedCountersInstance.countEvent('processing', 'broadcast_finalstate_noQueueEntry')
+            return
+          }
+
+          // validate corresponding tell sender
+          if (_sender == null) {
+            /* prettier-ignore */ if (logFlags.error) this.mainLogger.error(`poqo-data-and-receipt invalid sender for txid: ${payload.finalState.txid}, sender: ${_sender}`)
+            return
+          }
+
+          const isValidFinalDataSender = this.stateManager.transactionQueue.factValidateCorrespondingTellFinalDataSender(queueEntry, _sender)
+          if (isValidFinalDataSender === false) {
+            /* prettier-ignore */ if (logFlags.error) this.mainLogger.error(`poqo-data-and-receipt invalid sender ${_sender} for data: ${queueEntry.acceptedTx.txId}`)
+            return
+          }
+
+          if (!queueEntry.hasSentFinalReceipt) {
+            const executionGroupNodes = new Set(queueEntry.executionGroup.map(node => node.publicKey));
+            const hasTwoThirdsMajority = this.verifyAppliedReceipt(payload.receipt, executionGroupNodes)
```

Comment:
queueEntry is taken based on readableReq.finalState.txid, but verification is done based on readableReq.receipt.txid

Reply:
confirmed non binary endpoint is deleted here:
https://github.com/shardeum/shardus-core/pull/202/files
otherwise I would add the same error handling as mentioned in last note